File: low_structures.py
from collections import OrderedDict

import math
import cv2
import numpy as np


class orderless_memoized(object):
    def __init__(self, func):
        self.func = func
        self.cache = {}

    def __call__(self, *args):
        key = frozenset(args)
        try:
            value = self.cache[key]
        except KeyError:
            value = self.func(*args)
            self.cache[key] = value
        return value


def similarly_angled_lines(line_a, line_b):
    """
    lines has to differ by some value
    otherwise intersection is not interesting
    """
    th_a = line_a[1]
    th_b = line_b[1]

    return similar_angles(th_a, th_b)


@orderless_memoized
def similar_angles(angle_a, angle_b):
    min_ang_diff = 0.5  # ~28deg

    ang_diff = abs(angle_a - angle_b)
    if ang_diff < min_ang_diff or ang_diff > (np.pi - min_ang_diff):
        return True
    return False


def intersection(line_a, line_b):
    """
    Solve:
    x*cos(th_a) + y*sin(th_a) = r_a
    x*cos(th_b) + y*sin(th_b) = r_b

    As matrix:
    A*X = b
    """
    r_a, th_a = line_a
    r_b, th_b = line_b

    A = np.array([
        [np.cos(th_a), np.sin(th_a)],
        [np.cos(th_b), np.sin(th_b)],
    ])
    b = np.array([r_a, r_b])
    ok, point = cv2.solve(A, b)
    if ok:
        point = (point[0][0], point[1][0])
    return ok, point


def intersections(lines):
    points = {}

    for i, line_a in enumerate(lines):
        for j, line_b in enumerate(lines):
            if i <= j:
                continue

            ok, point = intersection(line_a, line_b)
            if not ok:
                continue

            key = tuple(sorted([i, j]))
            points[key] = point

    return points


def remove_duplicate_lines(lines, img_shape, min_dist=3):
    """
    duplicates: crosses in view at low angle
    """
    w, h = img_shape
    scope = 0.5
    min_x = 0 - w*scope
    min_y = 0 - h*scope
    max_x = w + w*scope
    max_y = h + h*scope

    to_remove = set([])
    for i, line_a in enumerate(lines):
        for j, line_b in enumerate(lines[i+1:], i+1):
            similar = similarly_angled_lines(line_a, line_b)
            if not similar:
                continue

            if abs(line_a[0] - line_b[0]) < min_dist:
                to_remove.add(max(i, j))
                continue

            ok, point = intersection(line_a, line_b)
            if not ok:
                continue

            x, y = point
            in_view = (
                min_x <= x <= max_x and
                min_y <= y <= max_y
            )

            if in_view:
                to_remove.add(max(i, j))

    cleaned = [line for i, line in enumerate(lines) if i not in to_remove]
    return cleaned


def generate_angle_buckets(bucket_size, step=None, ortogonal=True):
    """
    Creates a dict with ortogonal (if required) ranges for angles (in radians)
    Both bucket_size and step are taken in deg (it's easier to reason about)
    Angles between 0 and 180 deg

    Example output (bucket_size=20, step=5) - all values in deg:
    {
        45: [(35, 55), (125, 145)],
        50: [(40, 60), (130, 150)],
    }
    """

    window = np.deg2rad(bucket_size)
    step = np.deg2rad(step or bucket_size)

    window_2 = window/2.0
    pos = 0
    max_pos = (np.pi/2.0 if ortogonal else np.pi) - step
    buckets = OrderedDict()
    while 1:
        b1 = (pos - window_2, pos + window_2)
        bucket = [b1]

        if b1[0] < 0:
            b1_prim = (np.pi + b1[0], np.pi)
            bucket.append(b1_prim)

        if b1[1] > np.pi:
            b1_bis = (0, b1[1] - np.pi)
            bucket.append(b1_bis)

        if ortogonal:
            b2 = (b1[0] + np.pi/2, b1[1] + np.pi/2)
            bucket.append(b2)

            if b2[1] > np.pi:
                b2_prim = (0, b2[1] - np.pi)
                bucket.append(b2_prim)

        buckets[pos] = bucket

        pos += step
        if pos >= max_pos:
            break

    return buckets


def is_angle_in_bucket(angle, ranges):
    """
    takes a list of ranges for a given bucket,
    and angle to test
    """

    for start, end in ranges:
        if start <= angle <= end:
            return True

    return False


def lines_with_similar_angle(lines, angle):
    """
    Splits list of lines into one that are similar to given angle,
    and the rest of lines
    """

    similar = []
    other = []

    for line in lines:
        if similar_angles(line[1], angle):
            similar.append(line)
        else:
            other.append(line)

    return similar, other


def put_lines_into_buckets(buckets, lines):
    bucketed = []
    all_matches = set([])
    for bucket_angle, bucket in buckets.iteritems():
        matches = []

        for line in lines:
            if is_angle_in_bucket(line[1], bucket):
                matches.append(line)

        matches_key = tuple(matches)
        if matches and matches_key not in all_matches:
            all_matches.add(matches_key)
            bucketed.append((bucket_angle, matches))

    # sort by most numbers of matches in bucket
    bucketed = sorted(bucketed, key=lambda b: len(b[1]), reverse=True)
    return bucketed


def distance_between_points(point_a, point_b):
    return math.sqrt(
        (point_a[0]-point_b[0])**2 + (point_a[1]-point_b[1])**2
    )

File: low_structures.go
package sudoku

import (
	"fmt"
	"image"
	"math"

	"github.com/gonum/matrix/mat64"
)

const pi2 = 2 * math.Pi

type angleBucket struct {
	Start float64
	End   float64
}

func (b angleBucket) String() string {
	return fmt.Sprintf("Bucket{%.0f, %.0f}", b.Start*180/math.Pi, b.End*180/math.Pi)
}

type lineFragment struct {
	Start image.Point
	End   image.Point
}

func (f lineFragment) Length() float64 {
	return distanceBetweenPoints(f.Start, f.End)
}

func distanceBetweenPoints(a, b image.Point) float64 {
	return math.Hypot(float64(a.X-b.X), float64(a.Y-b.Y))
}

func similarAngles(a, b float64) bool {
	minAngDiff := 0.5 // ~28deg

	if a >= pi2 || a <= -pi2 {
		a -= math.Floor(a/pi2) * pi2
	}
	if b >= pi2 || b <= -pi2 {
		b -= math.Floor(b/pi2) * pi2
	}

	angDiff := math.Abs(a - b)
	if angDiff < minAngDiff || angDiff > (pi2-minAngDiff) {
		return true
	}
	return false
}

// Intersection calculates intersection of two lines
//
// Solve:
// x*cos(thetaA) + y*sin(thetaA) = rA
// x*cos(thetaB) + y*sin(thetaB) = rB
//
// As matrix:
// A*X = b
func intersection(lineA, lineB polarLine) (bool, image.Point) {
	A := mat64.NewDense(2, 2, []float64{
		math.Cos(lineA.Theta), math.Sin(lineA.Theta),
		math.Cos(lineB.Theta), math.Sin(lineB.Theta),
	})
	b := mat64.NewDense(2, 1, []float64{
		float64(lineA.Distance), float64(lineB.Distance),
	})
	x := mat64.NewDense(2, 1, nil)
	err := x.Solve(A, b)

	ok := err == nil
	// Using 0.5 to force round to nearest int rather than Floor
	point := image.Point{
		X: int(x.At(0, 0) + 0.5),
		Y: int(x.At(1, 0) + 0.5),
	}
	return ok, point
}

// duplicates: crosses in view at low angle
func removeDuplicateLines(lines []polarLine, width, height int) []polarLine {
	minDist := 3.0

	scope := 2
	minX := 0 - width/scope
	minY := 0 - height/scope
	maxX := width + width/scope
	maxY := height + height/scope

	toRemove := make(map[int]bool, len(lines))
	for i, lineA := range lines {
		for j, lineB := range lines[i+1:] {
			k := j + i + 1

			similar := similarAngles(lineA.Theta, lineB.Theta)
			if !similar {
				continue
			}

			if math.Abs(float64(lineA.Distance-lineB.Distance)) < minDist {
				toRemove[k] = true
				continue
			}

			ok, point := intersection(lineA, lineB)
			if !ok {
				continue
			}

			in_view := (minX <= point.X && point.X <= maxX &&
				minY <= point.Y && point.Y <= maxY)

			if in_view {
				toRemove[k] = true
				continue
			}
		}
	}

	deDuped := make([]polarLine, len(lines)-len(toRemove))
	j := 0
	for i, line := range lines {
		if toRemove[i] {
			continue
		}
		deDuped[j] = line
		j++
	}
	return deDuped
}

// generateAngleBuckets creates a map with ortogonal (if required) ranges for angles (in radians)
//     Both bucketSize and step are taken in deg (it's easier to reason about)
//     Angles between 0 and 180 deg
//
//     Example output (bucket_size=20, step=5) - all values in deg:
//     {
//         45: [(35, 55), (125, 145)],
//         50: [(40, 60), (130, 150)],
//     }
func generateAngleBuckets(bucketSize uint, step uint, ortogonal bool) map[float64][]angleBucket {
	const DegToRad float64 = math.Pi / 180

	window := DegToRad * float64(bucketSize)
	stepSize := DegToRad * float64(step)

	window2 := window / 2.0
	maxPos := math.Pi
	if ortogonal {
		maxPos = math.Pi / 2
	}
	maxPos = maxPos - stepSize
	buckets := make(map[float64][]angleBucket, 0)

	pos := 0.0
	for {
		b1 := angleBucket{pos - window2, pos + window2}
		bucket := []angleBucket{b1}

		if b1.Start < 0 {
			b1Prim := angleBucket{math.Pi + b1.Start, math.Pi}
			bucket = append(bucket, b1Prim)
		}

		if b1.End > math.Pi {
			b1Bis := angleBucket{0, b1.End - math.Pi}
			bucket = append(bucket, b1Bis)
		}

		if ortogonal {
			b2 := angleBucket{b1.Start + math.Pi/2, b1.End + math.Pi/2}
			bucket = append(bucket, b2)

			if b2.End > math.Pi {
				b2Prim := angleBucket{0, b2.End - math.Pi}
				bucket = append(bucket, b2Prim)
			}
		}

		buckets[pos] = bucket

		pos += stepSize
		if pos >= maxPos {
			break
		}
	}

	return buckets
}

// Splits lines into two groups one that are similar to given angle,
// and the rest of lines
func linesWithSimilarAngle(lines []polarLine, angle float64) ([]polarLine, []polarLine) {
	similar := make([]polarLine, 0)
	other := make([]polarLine, 0)

	for _, line := range lines {
		if similarAngles(line.Theta, angle) {
			similar = append(similar, line)
		} else {
			other = append(other, line)
		}
	}

	return similar, other
}

func putLinesIntoBuckets(buckets map[float64][]angleBucket, lines []polarLine) map[float64][]polarLine {
	bucketed := make(map[float64][]polarLine, 0)
	alreadyMatched := make(map[string]bool, 0)

	for angle, bucket := range buckets {
		matches := make([]polarLine, 0)

		for _, line := range lines {
			for _, b := range bucket {
				if b.Start <= line.Theta && line.Theta <= b.End {
					matches = append(matches, line)
					break
				}
			}
		}

		matchesKey := polarLineHash(matches).HashKey()
		if len(matches) > 0 && !alreadyMatched[matchesKey] {
			alreadyMatched[matchesKey] = true
			bucketed[angle] = matches
		}
	}
	return bucketed
}

// Bresenham's line algorithm
func pointsOnLineFragment(fragment lineFragment) []image.Point {
	x0, x1 := fragment.Start.X, fragment.End.X
	y0, y1 := fragment.Start.Y, fragment.End.Y

	dx := float64(x1 - x0)
	sx := 1
	if dx < 0 {
		sx = -1
		dx = -dx
	}

	dy := float64(y1 - y0)
	sy := 1
	if dy < 0 {
		sy = -1
		dy = -dy
	}

	points := make([]image.Point, 0)

	var err float64
	x, y := x0, y0
	if dx > dy {
		err = dx / 2.0
		for {
			points = append(points, image.Point{x, y})
			if x == x1 {
				break
			}

			err -= dy
			if err < 0 {
				y += sy
				err += dx
			}
			x += sx
		}
	} else {
		err = dy / 2.0
		for {
			points = append(points, image.Point{x, y})
			if y == y1 {
				break
			}

			err -= dx
			if err < 0 {
				x += sx
				err += dy
			}
			y += sy
		}
	}

	return points
}
